<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
  <head>
    

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">

 


      <title>Single Cell Level QC - </title>

  <meta name="description" content="Cell Level QC One thing that bothered me when I started working with 10X data (roughly 2016, after having worked with Smart-seq2 single nuclei data) is that for many metrics (percent intronic reads, percent intergenic reads, etc), though it was easy to get sample level values, but much less straightforward to get cell level metrics. Since then the situation has gotten better, with much work on using different metrics to help remove low quality cells.">
  <meta name="author" content="Sean Simmons"/><script type="application/ld+json">
{
    "@context": "http://schema.org",
    "@type": "WebSite",
    "name": "Is That Really What The Data Says?",
    
    "url": "https:\/\/seanken.github.io\/public"
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Organization",
  "name": "",
  "url": "https:\/\/seanken.github.io\/public"
  
  
  
  
}
</script>
<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [{
        "@type": "ListItem",
        "position": 1,
        "item": {
          "@id": "https:\/\/seanken.github.io\/public",
          "name": "home"
        }
    },{
        "@type": "ListItem",
        "position": 3,
        "item": {
          "@id": "https:\/\/seanken.github.io\/public\/posts\/single-cell-level-qc\/",
          "name": "Single cell level qc"
        }
    }]
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Article",
  "author": {
    "name" : "Sean Simmons"
  },
  "headline": "Single Cell Level QC",
  "description" : "Cell Level QC One thing that bothered me when I started working with 10X data (roughly 2016, after having worked with Smart-seq2 single nuclei data) is that for many metrics (percent intronic reads, percent intergenic reads, etc), though it was easy to get sample level values, but much less straightforward to get cell level metrics. Since then the situation has gotten better, with much work on using different metrics to help remove low quality cells.",
  "inLanguage" : "en",
  "wordCount":  916 ,
  "datePublished" : "2023-04-09T21:43:39",
  "dateModified" : "2023-04-09T21:43:39",
  "image" : "https:\/\/seanken.github.io\/public",
  "keywords" : [ "" ],
  "mainEntityOfPage" : "https:\/\/seanken.github.io\/public\/posts\/single-cell-level-qc\/",
  "publisher" : {
    "@type": "Organization",
    "name" : "https:\/\/seanken.github.io\/public",
    "logo" : {
        "@type" : "ImageObject",
        "url" : "https:\/\/seanken.github.io\/public",
        "height" :  60 ,
        "width" :  60
    }
  }
}
</script>

<meta property="og:title" content="Single Cell Level QC" />
<meta property="og:description" content="Cell Level QC One thing that bothered me when I started working with 10X data (roughly 2016, after having worked with Smart-seq2 single nuclei data) is that for many metrics (percent intronic reads, percent intergenic reads, etc), though it was easy to get sample level values, but much less straightforward to get cell level metrics. Since then the situation has gotten better, with much work on using different metrics to help remove low quality cells.">
<meta property="og:url" content="https://seanken.github.io/public/posts/single-cell-level-qc/" />
<meta property="og:type" content="website" />
<meta property="og:site_name" content="Is That Really What The Data Says?" />

  <meta name="twitter:title" content="Single Cell Level QC" />
  <meta name="twitter:description" content="Cell Level QC One thing that bothered me when I started working with 10X data (roughly 2016, after having worked with Smart-seq2 single nuclei data) is that for many metrics (percent intronic reads, â€¦">
  <meta name="twitter:card" content="summary_large_image" />
  <meta name="twitter:site" content="@seankenneths" />
  <meta name="twitter:creator" content="@seankenneths" />
  <meta name="generator" content="Hugo 0.111.3">
  <link rel="alternate" href="https://seanken.github.io/public/index.xml" type="application/rss+xml" title="Is That Really What The Data Says?"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css" integrity="sha384-AfEj0r4/OFrOo5t7NnNe46zW/tFgW6x/bCJG8FqQCEo3+Aro6EYUG4+cU+KJWu/X" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><link rel="stylesheet" href="https://seanken.github.io/public/css/main.css" /><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lora:400,700,400italic,700italic" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" />
  <link rel="stylesheet" href="https://seanken.github.io/public/css/highlight.min.css" /><link rel="stylesheet" href="https://seanken.github.io/public/css/codeblock.css" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.css" integrity="sha384-h/L2W9KefUClHWaty3SLE5F/qvc4djlyR4qY3NUV5HGQBBW7stbcfff1+I/vmsHh" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/default-skin/default-skin.min.css" integrity="sha384-iD0dNku6PYSIQLyfTOpB06F2KCZJAKLOThS5HRe8b3ibhdEQ6eKsFf/EeFxdOt5R" crossorigin="anonymous">


  </head>
  <body>
    <nav class="navbar navbar-default navbar-fixed-top navbar-custom">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#main-navbar">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="https://seanken.github.io/public">Is That Really What The Data Says?</a>
    </div>

    <div class="collapse navbar-collapse" id="main-navbar">
      <ul class="nav navbar-nav navbar-right">
        
          
            <li>
              <a title="Blog" href="/public">Blog</a>
            </li>
          
        
          
            <li>
              <a title="About" href="/public/page/about/">About</a>
            </li>
          
        

        

        
      </ul>
    </div>

    

  </div>
</nav>




    


<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>


  
  
  






  

  <header class="header-section ">
    
    
    <div class="intro-header no-img">
      <div class="container">
        <div class="row">
          <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
            <div class="posts-heading">
              
                <h1>Single Cell Level QC</h1>
              
              
                <hr class="small">
              
              
              
            </div>
          </div>
        </div>
      </div>
    </div>
  
  </header>


    
<div class="container" role="main">
  <div class="row">
    <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
      <article role="main" class="blog-post">
        <h2 id="cell-level-qc">Cell Level QC</h2>
<p>One thing that bothered me when I started working with 10X data (roughly 2016, after having worked with Smart-seq2 single nuclei data) is that for many metrics (percent intronic reads, percent intergenic reads, etc), though it was easy to get sample level values, but much less straightforward to get cell level metrics. Since then the situation has gotten better, with much work on using different metrics to help remove low quality cells. In particular, in recent years there has been the realization that % intronic reads in particular give a lot of insight for seperating real cells or nuclei from empty droplets and debris (the first such paper I am aware of was the DropletQC (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02547-0">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02547-0</a>) paper, with more recent papers from the Kellis lab (<a href="https://www.biorxiv.org/content/10.1101/2022.10.21.513315v1">https://www.biorxiv.org/content/10.1101/2022.10.21.513315v1</a>) looking at it for single nuclei data, or papers from the Linnarsson lab (<a href="https://www.biorxiv.org/content/10.1101/2022.10.24.513487v1">https://www.biorxiv.org/content/10.1101/2022.10.24.513487v1</a>) look at it at an atlas scale).</p>
<p>In particular, in the case of single nuclei data, the more reads that are from the nucleus the higher percent intronic reads one would expect, allowing us to seperate nuclei from empty droplets and other parts of the cell (something a recent paper we were slightly involved with explored to help understand ambient RNA (<a href="https://www.biorxiv.org/content/10.1101/2022.11.16.516780v1))">https://www.biorxiv.org/content/10.1101/2022.11.16.516780v1))</a>. This raises two questions: are there other metrics that will give us similiar (or even more) insight but which we arent reporting now? And are there better ways to get the information out after processing 10X data? In particular, for those who use CellRanger, though there is a lot of read level QC information hidden in the output bam, it can be hard to get at. In particular, to get at % intronic levels it is common to use other tools lik STARSolo, Kallisto, or others velocyto. Not to imply they are anything less than great tools (they are all great pieces of software!), but they do not calculate all the metrics we care about, and in many cases we might not want to use them.</p>
<p>Given the above, I decided it might be worth the time to write a script to extract as much cell level QC information at possible from the CellRanger bam. As such, I built a tool I am calling CellLevel_QC (<a href="https://github.com/seanken/CellLevel_QC)">https://github.com/seanken/CellLevel_QC)</a>. It is a simple java program, all one has to do is download the jar file and run it on the outs directory from CellRanger (see the github for details). The code does not do anything particularly clever, for the most part just goes line by line in the bam file and extracts information from each bam entry. It returns many cell level QC metrics that might be of interest (% intronic reads, % intergenic, % reads trimmed by CellRanger, etc).</p>
<h2 id="insight-from-cell-level-qc">Insight From Cell Level QC</h2>
<p>To explore this, I used the CellLevel_QC tool to look at one sample (12 week old mouse Hippocampus single nuclei sample taken from a recent prepub of ours, <a href="https://www.biorxiv.org/content/10.1101/2022.11.15.516665v1)">https://www.biorxiv.org/content/10.1101/2022.11.15.516665v1)</a>.</p>
<p>I started by running CellLevel_QC on the sample</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">java -jar SingleCellQC.jar -d outs -o out.txt
</span></span></code></pre></div><p>where <code>outs</code> is the output directory from CellRanger (v6.1.2, run with introns included with the mm10 reference). This results in an output file, <code>out.txt</code>, into R, and normalizing counts to percent:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">library(ggplot2)
</span></span><span class="line"><span class="cl">library(cowplot) //Not strictly needed, just to make look nice
</span></span><span class="line"><span class="cl">theme_set(theme_cowplot()) //Not strictly needed, just to make look nice
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">dat=read.table(&#34;out.txt&#34;,header=T)
</span></span><span class="line"><span class="cl">for(i in c(2:11,15)){dat[i]=100*dat[,i]/dat[,&#34;total&#34;]}dat=dat[dat$total&gt;10,] //Only look at droplets with at least 10 reads
</span></span><span class="line"><span class="cl">dat=dat[dat[,1]!=&#34;notCell&#34;,] //removes non-cell entry
</span></span><span class="line"><span class="cl">cells=scan(&#34;outs/filtered_feature_bc_matrix/barcodes.tsv.gz&#34;,&#34;&#34;) //Thinks cellranger thinks are nuclei
</span></span><span class="line"><span class="cl">dat[&#34;Cell&#34;]=dat$CBC %in% cells
</span></span></code></pre></div><p>The first thing to look at is the nUMI versus the % intronic reads, colored by if a droplet is declared a nuclei by CellRanger:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">ggplot(dat,aes(x=nUMI,y=intronic,color=Cell))+geom_point()+scale_x_log10()+ylab(&#34;Percent Intronic Reads&#34;)
</span></span></code></pre></div><p><img src="/Images_Single-Cell-Level-Q/Cell.label.pdf" alt="Identifying Nuclei"></p>
<p>Can see a nice seperation into 3 distinct clusters (with some other droplets here or there) as one might hope/expect, including seeing a cluster that corresponds very well to the CellRanger labelled nuclei (this is not always so clean, sometimes CellRanger definitely sets the cutoff poorly, eleading to loss of nuclei or empty droplets being misidentified as nuclei). We do, however, have many other metrics. We can first look at how the different metric compare using Spearman correlation:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">library(ComplexHeatmap)
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">COR=cor(dat[,c(2:14,16)],method=&#34;spearman&#34;)
</span></span><span class="line"><span class="cl">Heatmap(COR)
</span></span></code></pre></div><p>[insert plot]</p>
<p>For definition of each of the metrics see the CellLevel_QC github.</p>
<p>Can see obvious structure in this plot. The most obvious strucuture comes from total (total number of reads) being highly correlated with nUMI (not suprising), and from the percent antisense being highly correlated to percent intronic (also probably not suprising). One obvious question: can these other metrics be used to improve filtering? For example consider the polyA metric (the percent of reads in a cell that were trimmed to remove a polyA sequence). I particular, we can color the % intronic be nUMI plot by if the cell has &gt;1% of reads being trimmed for polyA:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">ggplot(dat,aes(x=nUMI,y=intronic,color=polyA&gt;1))+geom_point()+scale_x_log10()+ylab(&#34;Percent Intronic Reads&#34;)
</span></span></code></pre></div><p>[insert plot]</p>
<p>Can see that most of the droplets with high % polyA trimming are the lower nUMI, lower intronic droplets, which are likely empty droplets. We see something similiar for trimming of the TSO sequence, suggesting these metrics might be useful for better seperating empty droplets and real nuclei (in cases were it is less clear than the above).</p>
<h2 id="final-thoughts">Final Thoughts</h2>
<p>Anyways, just wanted to share a tool that we have found useful for some of our analysis&ndash;hope others find it useful as well! Wondering what other insight we can get from these metrics. Also: happy to think about adding other metrics as well (assuming I have the time), so let me know if there are any!</p>


        

        
            <hr/>
            <section id="social-share">
              <div class="list-inline footer-links">
                

<div class="share-box" aria-hidden="true">
    <ul class="share">
      
      <li>
        <a href="//twitter.com/share?url=https%3a%2f%2fseanken.github.io%2fpublic%2fposts%2fsingle-cell-level-qc%2f&amp;text=Single%20Cell%20Level%20QC&amp;via=seankenneths" target="_blank" title="Share on Twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.facebook.com/sharer/sharer.php?u=https%3a%2f%2fseanken.github.io%2fpublic%2fposts%2fsingle-cell-level-qc%2f" target="_blank" title="Share on Facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//reddit.com/submit?url=https%3a%2f%2fseanken.github.io%2fpublic%2fposts%2fsingle-cell-level-qc%2f&amp;title=Single%20Cell%20Level%20QC" target="_blank" title="Share on Reddit">
          <i class="fab fa-reddit"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.linkedin.com/shareArticle?url=https%3a%2f%2fseanken.github.io%2fpublic%2fposts%2fsingle-cell-level-qc%2f&amp;title=Single%20Cell%20Level%20QC" target="_blank" title="Share on LinkedIn">
          <i class="fab fa-linkedin"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.stumbleupon.com/submit?url=https%3a%2f%2fseanken.github.io%2fpublic%2fposts%2fsingle-cell-level-qc%2f&amp;title=Single%20Cell%20Level%20QC" target="_blank" title="Share on StumbleUpon">
          <i class="fab fa-stumbleupon"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.pinterest.com/pin/create/button/?url=https%3a%2f%2fseanken.github.io%2fpublic%2fposts%2fsingle-cell-level-qc%2f&amp;description=Single%20Cell%20Level%20QC" target="_blank" title="Share on Pinterest">
          <i class="fab fa-pinterest"></i>
        </a>
      </li>
    </ul>
  </div>
  

              </div>
            </section>
        

        
          

          
        
      </article>

      
        <ul class="pager blog-pager">
          
          
        </ul>
      


      
        
        
      

    </div>
  </div>
</div>

      
<footer>
  <div class="container">
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
        <ul class="list-inline text-center footer-links">
          
              <li>
		
		  <a href="mailto:seankenneths@gmail.com" title="Email me">
		
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fas fa-envelope fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
              <li>
		
		  <a href="https://github.com/seanken" title="GitHub">
		
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-github fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
              <li>
		
		  <a href="https://twitter.com/seankenneths" title="Twitter">
		
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-twitter fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
              <li>
		
		  <a href="https://linkedin.com/in/sean-simmons-8ba9a35a" title="LinkedIn">
		
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-linkedin fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
              <li>
		
		  <a rel="me"href="https://url" title="Mastodon">
		
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-mastodon fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
          
          <li>
            <a href="" title="RSS">
              <span class="fa-stack fa-lg">
                <i class="fas fa-circle fa-stack-2x"></i>
                <i class="fas fa-rss fa-stack-1x fa-inverse"></i>
              </span>
            </a>
          </li>
          
        </ul>
        <p class="credits copyright text-muted">
          
            
              Sean Simmons
            
          

          &nbsp;&bull;&nbsp;&copy;
          
            2023
          

          
            &nbsp;&bull;&nbsp;
            <a href="https://seanken.github.io/public">Is That Really What The Data Says?</a>
          
        </p>
        
        <p class="credits theme-by text-muted">
          <a href="https://gohugo.io">Hugo v0.111.3</a> powered &nbsp;&bull;&nbsp; Theme <a href="https://github.com/halogenica/beautifulhugo">Beautiful Hugo</a> adapted from <a href="https://deanattali.com/beautiful-jekyll/">Beautiful Jekyll</a>
          
        </p>
      </div>
    </div>
  </div>
</footer><script src="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.js" integrity="sha384-g7c+Jr9ZivxKLnZTDUhnkOnsh30B4H0rpLUpJ4jAIKs4fnJI+sEnkvrMWph2EDg4" crossorigin="anonymous"></script>
<script src="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/contrib/auto-render.min.js" integrity="sha384-mll67QQFJfxn0IYznZYonOWZ644AWYC+Pt2cHqMaRhXVrursRwvLnLaebdGIlYNa" crossorigin="anonymous"></script>
<script src="https://code.jquery.com/jquery-3.5.1.slim.min.js" integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj" crossorigin="anonymous"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<script src="https://seanken.github.io/public/js/main.js"></script>
<script src="https://seanken.github.io/public/js/highlight.min.js"></script>
<script> hljs.initHighlightingOnLoad(); </script>
<script> $(document).ready(function() {$("pre.chroma").css("padding","0");}); </script><script> renderMathInElement(document.body); </script><script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.js" integrity="sha384-QELNnmcmU8IR9ZAykt67vGr9/rZJdHbiWi64V88fCPaOohUlHCqUD/unNN0BXSqy" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe-ui-default.min.js" integrity="sha384-m67o7SkQ1ALzKZIFh4CiTA8tmadaujiTa9Vu+nqPSwDOqHrDmxLezTdFln8077+q" crossorigin="anonymous"></script><script src="https://seanken.github.io/public/js/load-photoswipe.js"></script>









    
  </body>
</html>

